Lib/tty.py

```diff
@@ -15,22 +24,93 @@
 OSPEED = 5
 CC = 6
 
-def setraw(fd, when=TCSAFLUSH):
-    """Put terminal into a raw mode."""
-    mode = tcgetattr(fd)
-    mode[IFLAG] = mode[IFLAG] & ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
-    mode[OFLAG] = mode[OFLAG] & ~(OPOST)
-    mode[CFLAG] = mode[CFLAG] & ~(CSIZE | PARENB)
-    mode[CFLAG] = mode[CFLAG] | CS8
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON | IEXTEN | ISIG)
+def mkecho(mode, echo=True):
+    """Set/unset ECHO."""
+    if echo:
+        mode[LFLAG] |= ECHO
+    else:
+        mode[LFLAG] &= ~ECHO
+
+def mkraw(mode):
+    """Raw mode termios."""
+    mode[IFLAG] &= ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
+    mode[OFLAG] &= ~(OPOST)
+    mode[CFLAG] &= ~(CSIZE | PARENB)
+    mode[CFLAG] |= CS8
+    mode[LFLAG] &= ~(ECHO | ICANON | IEXTEN | ISIG)
     mode[CC][VMIN] = 1
     mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+
+def setraw(fd, when=TCSAFLUSH):
+    """Put terminal into raw mode.
+    Returns original termios."""
+    mode = tcgetattr(fd)
+    new = list(mode)
+    mkraw(new)
+    tcsetattr(fd, when, new)
+    return mode
 
 def setcbreak(fd, when=TCSAFLUSH):
-    """Put terminal into a cbreak mode."""
+    """Put terminal into cbreak mode.
+    Returns original termios."""
     mode = tcgetattr(fd)
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON)
-    mode[CC][VMIN] = 1
-    mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+    new = list(mode)
+    new[LFLAG] &= ~(ECHO | ICANON)
+    new[CC][VMIN] = 1
+    new[CC][VTIME] = 0
+    tcsetattr(fd, when, new)
+    return mode
+
+def login_tty(fd):
+    """Makes the calling process a session leader; if fd is a
+    descriptor of a tty, then that tty becomes the stdin, stdout,
+    stderr, and controlling tty of the calling process. Closes fd."""
```

Comment:
This comment implies that fd can either be or not be a descriptor of a tty.
Not sure what would happen if it's not a tty descriptor.

Reply:
You are right. I will have to rephrase that. Actually, it will fail if `fd` is not a tty. *Do you think it would be a good idea to instead mention/document what exception would be raised if fd was not a tty?*